File: Level3/단어 변환.cpp
#include <string>
#include <vector>

using namespace std;

int result = 51; // 최소를 위한
vector<int> visited; // 방문리스트

// 최솟값 반환
int Min(int a, int b) {
    return a < b ? a : b;
}

// 1개만 차이가 나는지
bool is_trans_condi(string a, string b) {
    int cnt = 0;
    for (int i = 0; i < a.length(); ++i) {
        if (a[i] != b[i])
            cnt++;
    }
    
    return cnt == 1 ? 1 : 0;
}

// 깊이우선탐색
void DFS(vector<string>& words, int size, string cur, string target, int level) {
    // 탈출 조건
    if (cur == target) {
        result = Min(result, level);
        return;
    }
        
    for (int i = 0; i < size;++i) {
        // 조건
        if (is_trans_condi(cur, words[i]) == 1) {                  
            if (visited[i] == 1) // 재방문 방지
                continue;
        
            visited[i] = 1;
            DFS(words, size, words[i], target, level + 1);
            visited[i] = 0;
        }
    }
    return;
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    int size = words.size(); // 길이
    visited.resize(size, 0); // 초기화
    
    DFS(words, size, begin, target, 0);
    // 도달 가능하다면
    if (result != 51)
        answer = result;
    return answer;
}

File: Level3/단어 변환.py
#변환가능한지 검사
def check(cur, word):
    cnt=0
    for c,w in zip(cur,word):
        if c!=w:
            cnt+=1
        if cnt == 2:
            return False
    if cnt == 1:
        return True

#스택이용 깊이 우선 탐색
def DFS(begin, target, words, visited):
    cnt = 0;
    stack = [(begin,0)]

    #스택이 비어질때까지
    while stack:
        cur, dep = stack.pop()

        #탈출조건
        if cur == target:
            return dep

        for i in range(len(words)):
            if visited[i] == True:
                continue
            if check(cur,words[i]) == True:
                visited[i]=True
                stack.append((words[i],dep+1))
            
def solution(begin, target, words):
    answer = 0
    
    #타겟이 존재하지 않으면
    if target not in words:
        return 0
    
    #방문리스트 생성 및 초기화 
    visited = [False] * len(words)
    answer = DFS(begin,target,words,visited)

    return answer